src/vidpf.rs

```diff
@@ -0,0 +1,644 @@
+        y.conditional_negate(Choice::from(key.id));
+
+        let pi_i = &state.proof;
+        let cs_i = &public.cs.get(level).ok_or(VidpfError::IndexLevel)?;
+        let pi_tilde = Self::node_proof(input, level, &seed_new)?;
+        let h2_input = pi_tilde.conditional_xor(t_new, cs_i).xor(pi_i);
+        let pi_new = Self::node_proof_adjustment(h2_input).xor(pi_i);
+
+        let state_new = EvalState {
+            level: level + 1,
+            seed: seed_new,
+            t_control_bit: t_new,
+            proof: pi_new,
+        };
+
+        Ok((state_new, y))
+    }
+
+    /// [`Vidpf::open`] recovers the input's weight from additive shares produced
+    /// by aggregation servers.
+    ///
+    /// # Errors
+    /// Triggers an error when internal proofs are not equal.
+    pub fn open(a: ValueShare<Weight>, b: ValueShare<Weight>) -> Result<Weight, VidpfError> {
+        if a.proof != b.proof {
+            return Err(VidpfError::OpenSharesFailed);
+        }
+        Ok(a.share + b.share)
+    }
+
+    fn prg(&self, seed: &Seed) -> (Seed, Choice, Seed, Choice) {
+        let mut rng =
+            XofFixedKeyAes128::seed_stream(&XofSeed(*seed), DomainSepTag::PRNG, &self.nonce);
+
+        let mut sl = Seed::default();
+        let mut sr = Seed::default();
+        rng.fill_bytes(&mut sl);
+        rng.fill_bytes(&mut sr);
+        // Use the LSB of seeds as control bits -- This ensures [prg] costs two AES calls only.
+        let tl = Choice::from(sl[0] & 0x1);
+        let tr = Choice::from(sr[0] & 0x1);
+
+        (sl, tl, sr, tr)
+    }
+
+    fn convert(&self, seed: &Seed) -> (Seed, Weight) {
+        let mut rng =
+            XofFixedKeyAes128::seed_stream(&XofSeed(*seed), DomainSepTag::CONVERT, &self.nonce);
+
+        let mut out_seed = Seed::default();
+        rng.fill_bytes(&mut out_seed);
+        let value = Weight::generate(&mut rng, &self.weight_parameter);
+
+        (out_seed, value)
+    }
+
+    fn node_proof(input: &VidpfInput, level: usize, seed: &Seed) -> Result<Proof, VidpfError> {
+        let mut binder = input.prefix(level).to_bytes();
+        binder.extend_from_slice(
+            &u32::try_from(level)
+                .map_err(|_e| VidpfError::LevelTooBig)?
+                .to_be_bytes(),
+        );
+        let mut rng =
+            XofTurboShake128::seed_stream(&XofSeed(*seed), DomainSepTag::NODE_PROOF, &binder);
+
+        let mut proof = Proof::default();
+        rng.fill_bytes(&mut proof.0);
+
+        Ok(proof)
+    }
+
+    fn node_proof_adjustment(mut proof: Proof) -> Proof {
+        let mut rng = XofTurboShake128::seed_stream(
+            &XofSeed(Default::default()),
+            DomainSepTag::NODE_PROOF_ADJUST,
+            &proof.0,
+        );
+        rng.fill_bytes(&mut proof.0);
+
+        proof
+    }
+}
+
+/// Contains the domain separation tags for invoking different oracles.
+struct DomainSepTag;
+// [todo] Fix dst for the VIDPF protocol.
```

Comment:
Remove this and use the issue to track remaining work. (Per @tgeoghegan request: we have from time to time let TODOs in code rot.)

```suggestion
```